cranelift-codegen/src/legalizer/globalvalue.rs

```diff
@@ -123,7 +123,18 @@ fn load_addr(
 }
 
 /// Expand a `global_value` instruction for a symbolic name global.
-fn symbol(inst: ir::Inst, func: &mut ir::Function, gv: ir::GlobalValue, isa: &dyn TargetIsa) {
+fn symbol(
+    inst: ir::Inst,
+    func: &mut ir::Function,
+    gv: ir::GlobalValue,
+    isa: &dyn TargetIsa,
+    tls: bool,
+) {
     let ptr_ty = isa.pointer_type();
-    func.dfg.replace(inst).symbol_value(ptr_ty, gv);
+
+    if tls {
+        func.dfg.replace(inst).tls_value(ptr_ty, gv);
+    } else {
+        func.dfg.replace(inst).symbol_value(ptr_ty, gv);
+    }
```

Comment:
Instead of having a `tls_value` instruction, would it work to just continue to use `symbol_value` here, and use the `tls` flag of `gv` to determine whether it's TLS or not? You could add a `is_tls_data` predicate function which could work the same way as `is_colocated_data`, `FormatPredicateKind::IsColocatedData`, and so on, and then `define_entity_ref` in cranelift-codegen/meta/src/isa/x86/encodings.rs could use that to decide when to apply the TLS legalizations.

Reply:
`symbol_value` can directly be encoded, while `tls_value` needs to be legalized to the correct instruction for the respective object format. See comment below for why. As far as I know, unlike an encoding, there is no way to predicate a legalization.